File: sandbox-runner/patterns/fanout-child-workflows/python/workflows.py
import asyncio
from datetime import timedelta

from temporalio import workflow
from temporalio.workflow import ChildWorkflowHandle

from activities import process_record
from shared import CHUNK_SIZE, TASK_QUEUE


@workflow.defn
class RecordBatchWorkflow:
    """Child workflow: processes a contiguous slice of records [offset, offset+length)."""

    @workflow.run
    async def run(self, offset: int, length: int) -> int:
        processed = 0
        for i in range(offset, offset + length):
            await workflow.execute_activity(
                process_record,
                i,
                start_to_close_timeout=timedelta(seconds=10),
            )
            processed += 1
        workflow.logger.info(f"Batch complete: offset={offset} length={length} processed={processed}")
        return processed


@workflow.defn
class FanOutWorkflow:
    """Parent workflow: splits the total record set into chunks and fans out to
    one child workflow per chunk."""

    @workflow.run
    async def run(self, total_records: int, chunk_size: int = CHUNK_SIZE) -> int:
        parent_id = workflow.info().workflow_id
        handles: list[ChildWorkflowHandle] = []

        offset = 0
        while offset < total_records:
            length = min(chunk_size, total_records - offset)
            handle = await workflow.start_child_workflow(
                RecordBatchWorkflow.run,
                args=[offset, length],
                id=f"{parent_id}/batch-{offset}",
                task_queue=TASK_QUEUE,
            )
            handles.append(handle)
            offset += chunk_size

        results = await asyncio.gather(*handles)
        total = sum(results)
        workflow.logger.info(
            f"Fan-out complete: totalRecords={total_records} chunks={len(handles)} total={total}"
        )
        return total

File: sandbox-runner/patterns/fanout-child-workflows/typescript/workflows.ts
import {
  executeChild,
  log,
  proxyActivities,
  workflowInfo,
} from "@temporalio/workflow";

import type * as activities from "./activities";
import { CHUNK_SIZE, TASK_QUEUE } from "./shared";

const { processRecord } = proxyActivities<typeof activities>({
  startToCloseTimeout: "10 seconds",
});

/**
 * Child workflow: processes a contiguous slice of records [offset, offset+length).
 */
export async function recordBatchWorkflow(
  offset: number,
  length: number
): Promise<number> {
  let processed = 0;
  for (let i = offset; i < offset + length; i++) {
    await processRecord(i);
    processed++;
  }
  log.info(`Batch complete`, { offset, length, processed });
  return processed;
}

/**
 * Parent workflow: splits the total record set into chunks and fans out to
 * one child workflow per chunk.
 */
export async function fanOutWorkflow(
  totalRecords: number,
  chunkSize: number = CHUNK_SIZE
): Promise<number> {
  const parentId = workflowInfo().workflowId;
  const children: Promise<number>[] = [];

  for (let offset = 0; offset < totalRecords; offset += chunkSize) {
    const length = Math.min(chunkSize, totalRecords - offset);
    children.push(
      executeChild(recordBatchWorkflow, {
        args: [offset, length],
        taskQueue: TASK_QUEUE,
        workflowId: `${parentId}/batch-${offset}`,
      })
    );
  }

  const results = await Promise.all(children);
  const total = results.reduce((sum, n) => sum + n, 0);
  log.info(`Fan-out complete`, { totalRecords, chunks: children.length, total });
  return total;
}
